dockers/docker-base/etc/supervisor/kill_supervisor.py

```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/python
+
+# Please follow the link for documentation: http://supervisord.org/events.html
+# SupervisorD exits immediately if one of its managed process get crashed as it subscribes to "EVENT Listener" process.
+
+import sys
+import os
+import signal
+import subprocess
+
+from supervisor.childutils import listener
+
+def write_stdout(s):
+    # only eventlistener protocol messages may be sent to stdout
+    sys.stdout.write(s)
+    sys.stdout.flush()
+
+def write_stderr(s):
+    sys.stderr.write(s)
+    sys.stderr.flush()
+
+def main():
+    while True:
+        all_service_list = []
+        proc = subprocess.Popen(["supervisorctl avail | cut -d' ' -f1"], shell=True, stdout=subprocess.PIPE)
+        (out, err) = proc.communicate()
+
+        all_service_list = out.split()
+
+        # "exception_service_list" contains all the program excluded  from event listener process.
+        exception_service_list = ["start.sh", "enable_counters", "swssconfig", "arp_update", "ledinit", "fancontrol", "lm-sensors", "ledd", "xcvrd", "configdb-load.sh", "snmpd-config-updater"]
```

Comment:
Listing all excluded processes (which reside in multiple Docker containers) together in docker-base breaks the concept of keeping each container compartmentalized and separate from one another.

I was recently working on a supervisor event listener that performs a similar task (kill supervisor if a supervised process exits abnormally). My intention was to implement this on a container-by-container basis, as needed. However, to implement something like this in all containers would result in a lot of duplicate code. I can understand why you would want to only have one copy of the logic, but I feel like it would be better to have container-specific information (i.e., the list of excluded processes) specified in the container where the processes reside.

Maybe there's a way to have the script with the logic live in docker-base, and pass the excluded processes from within each container?

Reply:
this list should be per docker configuration, not on the docker base.

Reply:
@nikos-github: If supervisor (which is PID 1 inside the Docker container) is killed, the container will stop, at which point, the container can be restarted gracefully by systemd, thus attempting an automated mitigation instead of keeping the container running in a potentially broken state.

Reply:
@nikos-github: The concept here is if a process inside a container exits abnormally, most likely the container is no longer functioning as expected. So, instead of keeping the container running in a bad state, this feature would kill supervisor, thus causing the container to stop and get restarted gracefully.

Reply:
I have my doubts and objections as to why we would need to do that to supervisord and whether this will actually intervene, unintentionally, with warm restart functionality. In general this may be an undesirable thing to do - it's quite a big hammer to use and say that if the process died, then have the supervisor die too in order for the container to restart. What if I had some other important custom processes running inside? What if it disrupts (and it can), GR? I think at minimum the behavior should be configurable and not on by default. Have we considered not killing the supervisor but the rest of the processes that it controls and have those restart. The event listener hook can run arbitrary code anyway.

Reply:
I think nikos have a valid point here. the behavior should be customizable in someway. For example, in team docker, there are multiple independent teamd process, and each controls one lag. assuming one teamd crashes, we probably do not want to crash the whole teamd docker and restart again. We probably should just restart the teamd process.